### PR TITLE
MangaHub Factory: get Edo Tensei

### PR DESCRIPTION
**Pull request by Riztard**
- back to scrapping
- add user agent randomizer(latest chrome)
- make css selector more general
- use new pageListParse method
- add back some source: MangaHub.io, MangaPanda.onl, MangaReader.site

Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [ ] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

**Comment by vetleledaal** [comment]
Does switching back to scraping fix the cookie issue, which prompted the removal in the first place? Test with some single chapter mange like `Tomo-chan wa Onnanoko` and confirm you can load 11 chapters within ~20 minutes.

Also update the issue moderator and removed sources list as appropriate.

**Comment by Riztard** [comment]
nvm i just realize i have another issue with the chapter page only loading max 6

**Comment by Riztard** [comment]
> Does switching back to scraping fix the cookie issue, which prompted the removal in the first place? Test with some single chapter mange like `Tomo-chan wa Onnanoko` and confirm you can load 11 chapters within ~20 minutes.

not sure how you test that. read 1 by 1 very slow?

> Also update the issue moderator and removed sources list as appropriate.

just leave it first just in case they broke it again intentionaly

**Comment by vetleledaal** [comment]
>> confirm you can load 11 chapters within ~20 minutes.
>
> not sure how you test that. read 1 by 1 very slow?

Opposite, read 11 chapters fast. Within (less than) 20 minutes so you hit any rate limiting they put in. Should be doable in 1-2 minutes with manga that has a single page for each chapter.

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt [hunk 1]
```diff
@@ -251,11 +251,82 @@ abstract class MangaHub(
...
+
+            for (page in 1..maxPage) {
+                val url = urlTemplate.replaceAfterLast("/", "$page.$extension")
+                val pageObject = Page(page - 1, "", url)
+                pages.add(pageObject)
+            }
+        } else {
+
+            // if there is no max pages string make some calls to check if the pages exist
+            // increase or decreasing in big number to small to reduce call[ 50 25 10 4 1] if maxPages is 20+ ~200+
```
Comment (vetleledaal) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt]: The pattern you're looking for is binary search. That gives you the answer in O(log n) requests on average

**Comment by Riztard** [comment]
> > not sure how you test that. read 1 by 1 very slow?
> 
> Opposite, read 11 chapters fast. Within (less than) 20 minutes so you hit any rate limiting they put in. Should be doable in 1-2 minutes with manga that has a single page for each chapter.

then it works been through 30 ch. maybe the rate limit  only for API

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt [hunk 2]
```diff
@@ -268,54 +292,10 @@ abstract class MangaHub(
...
-                            }
-                        }
-                    }
-                    break
-                }
-            }
+            // if there is no max pages string make some calls to check if the pages exist using findPageCount()
+            // increase or decreasing by using binary search algorithm
+            val headers = headersBuilder().build()
+            val maxPage = findPageCount(urlTemplate, extension, client, headers)
```
Comment (vetleledaal) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt:298]: you don't need to send in `client` or `headers`. should already be set in the class

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt [hunk 3]
```diff
@@ -251,6 +251,30 @@ abstract class MangaHub(
...
+
+        while (lowerBound <= upperBound) {
+            val midpoint = lowerBound + (upperBound - lowerBound) / 2
+            val url = urlTemplate.replaceAfterLast("/", "$midpoint.$extension")
+            val request = Request.Builder()
+                .url(url)
+                .headers(headers)
+                .head()
+                .build()
+            val response = client.newCall(request).execute()
```
Comment (vetleledaal) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt:266]: maybe wrap the call in a try/catch. So you can either show a better error to the user, or on error assume the image doesn't exist.

Comment (vetleledaal) [reply multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt:266]:
```suggestion
            val response = try {
                client.newCall(request).execute()
            } catch (e: IOException) {
                throw Exception("Failed to fetch $url")
            }
```

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt [hunk 4]
```diff
@@ -0,0 +1,416 @@
+package eu.kanade.tachiyomi.multisrc.mangahub
+
+import android.util.Log
```
Comment (vetleledaal) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt:3]: Unused import 😤

```suggestion
```

Comment (Riztard) [reply multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt:3]: ah forgot to remove that lol

**Comment by alessandrojean** [comment]
Is this done for merge?

**Comment by Riztard** [comment]
> Is this done for merge?

it should be, unless i/we miss something. atleast it works

**Pull request merged**